### ISSUE TITLE
Inconsistency when typechecking dunders

### ISSUE THREAD

**Issue by mniip** [microsoft/pyright#3502]
**Describe the bug**
CPython will look up dunder methods among any attributes that are implemented by descriptors. In the standard case a dunder method would be implemented by a `def` within the class body, so the function object is acting as a descriptor. However it is also possible to use e.g. `__slots__`, in which case a `member_descriptor` is used.

From the above it follows that the following thing will fail:
```python
class C:
    def __init__(self):
        self.__iter__ = lambda: iter([])
        self.__add__ = lambda x: x

for _ in C(): pass # TypeError: 'C' object is not iterable
C() + C() # TypeError: unsupported operand type(s) for +: 'C' and 'C'
```
However the following thing will succeed:
```python
class D:
    __slots__ = "__iter__", "__add__"
    def __init__(self):
        self.__iter__ = lambda: iter([])
        self.__add__ = lambda x: x

for _ in D(): pass
D() + D()
```

**To Reproduce**
Running pyright on both snippets reveals that pyright thinks that both types are addable but neither is iterable:
```
error: "C" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
error: "D" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
```
Pyright makes no distinction between `C` and `D` (what actually determines whether it crashes), however also decides that `__add__` is fine but `__iter__` is not.

**Expected behavior**
In an ideal world pyright would consider `D` as both addable and iterable, and `C` as neither addable nor iterable. However I am not sure how difficult it is for pyright to work with general descriptors.
Maybe a special case for `property` and `__slots__`?
In any case at least there should not be a discrepancy between `__add__` and `__iter__` (I haven't really tested all other dunders).

**Screenshots or Code**

**VS Code extension or command-line**
commandline pyright 1.1.248

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release. There were two separate bugs here. The first accounted for the inconsistency between `__iter__` and `__add__`. The second affected the handling of `__slots__`. Both are now fixed.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.250, which I just published. It will also be included in the next release of pylance.
